In [1]:
!git init
!git remote -v

Reinitialized existing Git repository in /home/Priya_darshan/GROUP_01_BE623_project1/.git/
origin	https://github.com/agrimagrawalbiotech-crypto/repo-repo.git (fetch)
origin	https://github.com/agrimagrawalbiotech-crypto/repo-repo.git (push)


1 (a)

In [2]:
import requests

base = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi"
params = {
    "db": "nucleotide",
    "id": "NC_012920.1",
    "rettype": "fasta",
    "retmode": "text",
}
r = requests.get(base, params=params)
print(r.url)
print(r.text[0:50])

https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=NC_012920.1&rettype=fasta&retmode=text
>NC_012920.1 Homo sapiens mitochondrion, complete 


In [16]:
with open("data/"NC_012920.1.fasta", "w") as fh:
    fh.write(r.text)

fatal: pathspec 'data/NC_012920.1.fasta' did not match any files
On branch master
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   Project1-Priyadarshan.ipynb
	deleted:    Project1.ipynb

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.ipynb_checkpoints/
	NC_012920.1.fasta
	NC_012920.1.gb
	YP_003024029.1.fasta
	deposited_protein.txt
	headers.txt
	rna_sequence.txt
	sequence.txt

no changes added to commit (use "git add" and/or "git commit -a")


Tools used: Python (requests, dictionary, open(), and write()) and NCBI E-utilities API.

1 (b)

In [4]:
import requests

base = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi"
params = {
    "db": "nucleotide",
    "id": "NC_012920.1",
    "rettype": "gb",
    "retmode": "text",
}
r = requests.get(base, params=params)
print(r.url)
print(r.text[0:100])

https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=NC_012920.1&rettype=gb&retmode=text
LOCUS       NC_012920              16569 bp    DNA     circular PRI 03-APR-2023
DEFINITION  Homo sap


In [5]:
with open("NC_012920.1.gb", "w") as fh:
    fh.write(r.text)

ACCESSION   NC_012920 AC_000021
VERSION     NC_012920.1
ORGANISM  Homo sapiens
LOCUS       NC_012920              16569 bp
gene            7586..8269
                /gene="COX2"
                /gene_synonym="COII; MTCO2"
CDS coordinates  7586..8269
protein_id="YP_003024029.1"

Tools used: Python (requests, dictionary, open(), and write()) and NCBI E-utilities API

1 (c)

In [6]:
import requests

base = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi"
params = {
    "db": "protein",
    "id": "YP_003024029.1",
    "rettype": "fasta",
    "retmode": "text",
}
r = requests.get(base, params=params)
print(r.url)
print(r.text)

https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&id=YP_003024029.1&rettype=fasta&retmode=text
>YP_003024029.1 cytochrome c oxidase subunit II (mitochondrion) [Homo sapiens]
MAHAAQVGLQDATSPIMEELITFHDHALMIIFLICFLVLYALFLTLTTKLTNTNISDAQEMETVWTILPA
IILVLIALPSLRILYMTDEVNDPSLTIKSIGHQWYWTYEYTDYGGLIFNSYMLPPLFLEPGDLRLLDVDN
RVVLPIEAPIRMMITSQDVLHSWAVPTLGLKTDAIPGRLNQTTFTATRPGVYYGQCSEICGANHSFMPIV
LELIPLKIFEMGPVFTL




In [7]:
with open("YP_003024029.1.fasta", "w") as fh:
    fh.write(r.text)

In [8]:
!grep ">" YP_003024029.1.fasta
!grep -v ">" YP_003024029.1.fasta | tr -d '\n' | wc -c

>YP_003024029.1 cytochrome c oxidase subunit II (mitochondrion) [Homo sapiens]
227


Tools used: Python (requests, dictionary, open(), and write()), NCBI E-utilities API, and Ubuntu shell commands (grep, tr, and wc).

1(d)

In [9]:
!grep ">" NC_012920.1.fasta YP_003024029.1.fasta > headers.txt

In [10]:
!cat headers.txt

NC_012920.1.fasta:>NC_012920.1 Homo sapiens mitochondrion, complete genome
YP_003024029.1.fasta:>YP_003024029.1 cytochrome c oxidase subunit II (mitochondrion) [Homo sapiens]


Tools used: Ubuntu shell commands (grep and cat).

2 (a)

In [11]:
!grep -v ">" NC_012920.1.fasta | tr -d '\n' > sequence.txt
!sed '/^>/!s/T/U/g' sequence.txt > rna_sequence.txt

sequence = ""

for line in open("rna_sequence.txt"):
    sequence = sequence + line
cds = sequence[7585:8269]
print("length in nucleotides =", len(cds))
print("remainder when divided by 3 =", len(cds) % 3)

length in nucleotides = 684
remainder when divided by 3 = 0


Tools used: Ubuntu shell commands (grep, tr, and sed) and Python (open(), for loop, string slicing, len(), and modulo operator %).

2 (b)

In [12]:
genetic_code = {
    "AAA": "K", "AAU": "N", "AAG": "K", "AAC": "N",
    "AUA": "M", "AUU": "I", "AUG": "M", "AUC": "I",
    "AGA": "",  "AGU": "S", "AGG": "",  "AGC": "S",
    "ACA": "T", "ACU": "T", "ACG": "T", "ACC": "T",
    "UAA": "",  "UAU": "Y", "UAG": "",  "UAC": "Y",
    "UUA": "L", "UUU": "F", "UUG": "L", "UUC": "F",
    "UGA": "W", "UGU": "C", "UGG": "W", "UGC": "C",
    "UCA": "S", "UCU": "S", "UCG": "S", "UCC": "S",
    "GAA": "E", "GAU": "D", "GAG": "E", "GAC": "D",
    "GUA": "V", "GUU": "V", "GUG": "V", "GUC": "V",
    "GGA": "G", "GGU": "G", "GGG": "G", "GGC": "G",
    "GCA": "A", "GCU": "A", "GCG": "A", "GCC": "A",
    "CAA": "Q", "CAU": "H", "CAG": "Q", "CAC": "H",
    "CUA": "L", "CUU": "L", "CUG": "L", "CUC": "L",
    "CGA": "R", "CGU": "R", "CGG": "R", "CGC": "R",
    "CCA": "P", "CCU": "P", "CCG": "P", "CCC": "P"
}

protein = ""

for i in range(0, len(cds), 3):
    codon = cds[i:i+3]
    protein = protein + genetic_code[codon]
print("First 30 residues =", protein[:30])
print("Total protein length =", len(protein))

First 30 residues = MAHAAQVGLQDATSPIMEELITFHDHALMI
Total protein length = 227


I used the mitochondrial RNA genetic code for self-translation. Reference: NCBI translation table 2 (transl_table=2) Tools used: Python dictionary, for loop, range(), string slicing, string concatenation, and len(). The mitochondrial RNA genetic code was based on NCBI transl_table=2.

2 (c)

In [13]:
!grep -v ">" YP_003024029.1.fasta | tr -d '\n' > deposited_protein.txt

deposited_protein = ""

for line in open("deposited_protein.txt"):
    deposited_protein = deposited_protein + line
print(protein == deposited_protein)

True


The translated protein sequence and the deposited protein sequence are identical.
Tools used: Ubuntu shell commands (grep and tr) and Python (open(), for loop, string concatenation, and string comparison using ==).

2 (d)

In [14]:
from Bio.Seq import Seq

biopython_protein = Seq(cds).translate(table=2)

print("My translation:", protein)
print("Biopython translation:", biopython_protein)
print("Do they agree:", protein == biopython_protein)

My translation: MAHAAQVGLQDATSPIMEELITFHDHALMIIFLICFLVLYALFLTLTTKLTNTNISDAQEMETVWTILPAIILVLIALPSLRILYMTDEVNDPSLTIKSIGHQWYWTYEYTDYGGLIFNSYMLPPLFLEPGDLRLLDVDNRVVLPIEAPIRMMITSQDVLHSWAVPTLGLKTDAIPGRLNQTTFTATRPGVYYGQCSEICGANHSFMPIVLELIPLKIFEMGPVFTL
Biopython translation: MAHAAQVGLQDATSPIMEELITFHDHALMIIFLICFLVLYALFLTLTTKLTNTNISDAQEMETVWTILPAIILVLIALPSLRILYMTDEVNDPSLTIKSIGHQWYWTYEYTDYGGLIFNSYMLPPLFLEPGDLRLLDVDNRVVLPIEAPIRMMITSQDVLHSWAVPTLGLKTDAIPGRLNQTTFTATRPGVYYGQCSEICGANHSFMPIVLELIPLKIFEMGPVFTL*
Do they agree: False


The direct comparison between my translation and Biopython's translation returned False. This is because my manually constructed genetic code dictionary represents stop codons as an empty string (""), whereas Biopython represents the terminal stop codon as *. Therefore, the amino acid residues are the same, but the two protein sequence strings are not identical because of the different representation of the terminal stop codon.

Tools used: Biopython (Bio.Seq.Seq and translate()), Python string comparison (==), and the vertebrate mitochondrial genetic code (translation table 2).

3 (a)

In [15]:
translated_length = len(protein)
deposited_length = len(deposited_protein)

print("Translated length:", translated_length)
print("Deposited length:", deposited_length)
print("Difference:", translated_length - deposited_length)

Translated length: 227
Deposited length: 227
Difference: 0


Tools used: Python len() function, variables, and arithmetic subtraction (-).

3 (c)

There is a difference between deposited protein and direct translation using the standard genetic code because MT-CO2 is encoded by mitochondrial DNA which uses the vertebrate mitochondrial genetic code (translation table 2). So, there are differences in codons: AUA codes for methionine(vertebrate mitochondiral) instead of isoleucine(standard), UGA codes for tryptophan instead of being a stop codon(standard), and AGA and AGG are stop codons instead of coding for arginine(standard). Thus, these differences in codons can change the resulting protein sequence when the mitochondrial sequence is translated through standard genetic code.
Agrim = The protein sequence deoes not match because it is a selenoprotein which has a selenocysteine residue. selenocysteine is not naturaly included in the standard genetic code and is misrepresented as * when translated. The * means a stop codon but it is not the case, instead it should be U for selenocysteine. If updated genetic code is used then there is no difference between deposited and translated sequences. 

Source of information:- https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes#SG2